pkgs/os-specific/linux/systemd/default.nix

```diff
@@ -46,79 +216,39 @@ in stdenv.mkDerivation rec {
-    [ libffi audit lz4 bzip2 libapparmor
-      iptables gnu-efi
-      # This is actually native, but we already pull it from buildPackages
-      pythonLxmlEnv
-    ] ++ stdenv.lib.optional withSelinux libselinux;
+    [ linuxHeaders libcap libuuid libffi ] ++
+    getDeps featureBuildInputs ++
+    stdenv.lib.optional withKexectools kexectools;
 
   #dontAddPrefix = true;
 
-  mesonFlags = [
-    "-Dloadkeys-path=${kbd}/bin/loadkeys"
-    "-Dsetfont-path=${kbd}/bin/setfont"
-    "-Dtty-gid=3" # tty in NixOS has gid 3
-    # "-Dtests=" # TODO
-    "-Dlz4=true"
-    "-Dhostnamed=true"
-    "-Dnetworkd=true"
-    "-Dsysusers=false"
-    "-Dtimedated=true"
-    "-Dtimesyncd=true"
-    "-Dfirstboot=false"
-    "-Dlocaled=true"
-    "-Dresolve=true"
-    "-Dsplit-usr=false"
-    "-Dlibcurl=false"
-    "-Dlibidn=false"
-    "-Dlibidn2=true"
-    "-Dquotacheck=false"
-    "-Dldconfig=false"
-    "-Dsmack=true"
-    "-Dsystem-uid-max=499" #TODO: debug why awking around in /etc/login.defs doesn't work
-    "-Dsystem-gid-max=499"
-    # "-Dtime-epoch=1"
-
-    (if stdenv.isAarch32 || stdenv.isAarch64 || !stdenv.hostPlatform.isEfi then "-Dgnu-efi=false" else "-Dgnu-efi=true")
-    "-Defi-libdir=${toString gnu-efi}/lib"
-    "-Defi-includedir=${toString gnu-efi}/include/efi"
-    "-Defi-ldsdir=${toString gnu-efi}/lib"
-
-    "-Dsysvinit-path="
-    "-Dsysvrcnd-path="
-
-    "-Dkill-path=${coreutils}/bin/kill"
-    "-Dkmod-path=${kmod}/bin/kmod"
-    "-Dsulogin-path=${utillinux}/bin/sulogin"
-    "-Dmount-path=${utillinux}/bin/mount"
-    "-Dumount-path=${utillinux}/bin/umount"
-  ];
-
+  # Passing either mesonFlags or mesonFlagsArray via the environment makes it
+  # very difficult to preserve flags which might contain whitespace or shell
+  # special characters, so we use bash syntax to set mesonFlagsArray instead.
   preConfigure = ''
-    mesonFlagsArray+=(-Dntp-servers="0.nixos.pool.ntp.org 1.nixos.pool.ntp.org 2.nixos.pool.ntp.org 3.nixos.pool.ntp.org")
-    mesonFlagsArray+=(-Ddbuspolicydir=$out/etc/dbus-1/system.d)
-    mesonFlagsArray+=(-Ddbussessionservicedir=$out/share/dbus-1/services)
-    mesonFlagsArray+=(-Ddbussystemservicedir=$out/share/dbus-1/system-services)
-    mesonFlagsArray+=(-Dpamconfdir=$out/etc/pam.d)
-    mesonFlagsArray+=(-Drootprefix=$out)
-    mesonFlagsArray+=(-Drootlibdir=$lib/lib)
-    mesonFlagsArray+=(-Dpkgconfiglibdir=$dev/lib/pkgconfig)
-    mesonFlagsArray+=(-Dpkgconfigdatadir=$dev/share/pkgconfig)
+    mesonFlagsArray+=(${lib.concatStringsSep " " [
+      (lib.escapeShellArgs (lib.mapAttrsToList featureFlag finalFeatures))
+      "-Ddbuspolicydir=$out/etc/dbus-1/system.d"
+      "-Ddbussessionservicedir=$out/share/dbus-1/services"
+      "-Ddbussystemservicedir=$out/share/dbus-1/system-services"
+      "-Dpamconfdir=$out/etc/pam.d"
+      "-Dpkgconfigdatadir=$dev/share/pkgconfig"
+      "-Dpkgconfiglibdir=$dev/lib/pkgconfig"
+      "-Drootlibdir=$lib/lib"
+      "-Drootprefix=$out"
```

Comment:
using placeholder of nix2 this could be moved to `mesonFlags` in order to avoid escaping.

Reply:
Thanks for reviewing! I don't know what that means though. Can you give an example or point me at the right documentation?

Reply:
Derivation attributes can now reference the outputs of the derivation using the placeholder builtin function. For example, the attribute
```
configureFlags = "--prefix=${placeholder "out"} --includedir=${placeholder "dev"}";
```
will cause the configureFlags environment variable to contain the actual store paths corresponding to the out and dev outputs.